icu4c/source/common/cmemory.h

```diff
@@ -675,6 +677,77 @@ inline H *MaybeStackHeaderAndArray<H, T, stackCapacity>::orphanOrClone(int32_t l
     return p;
 }
 
+/**
+ * A simple memory management class that creates new heap allocated objects (of
+ * any class that has a public constructor), keeps track of them and eventually
+ * deletes them all in its own destructor.
+ *
+ * A typical use-case would be code like this:
+ *
+ *     MemoryPool<MyType> pool;
+ *
+ *     MyType* o1 = pool.create();
+ *     if (o1 != nullptr) {
+ *         foo(o1);
+ *     }
+ *
+ *     MyType* o2 = pool.create(1, 2, 3);
+ *     if (o2 != nullptr) {
+ *         bar(o2);
+ *     }
+ *
+ *     // MemoryPool will take care of deleting the MyType objects.
+ *
+ * It doesn't do anything more than that, and is intentionally kept minimalist.
+ */
+template<typename T>
+class MemoryPool {
+public:
+    MemoryPool() : count(0), pool() {}
+
+    ~MemoryPool() {
+        for (int32_t i = 0; i < count; ++i) {
+            delete pool[i];
+        }
+    }
+
+    MemoryPool(const MemoryPool&) = delete;
+    MemoryPool& operator=(const MemoryPool&) = delete;
+
+    MemoryPool(MemoryPool&& other) U_NOEXCEPT : count(other.count),
+                                                pool(std::move(other.pool)) {
+        other.count = 0;
+    }
+
+    MemoryPool& operator=(MemoryPool&& other) U_NOEXCEPT {
+        count = other.count;
+        pool = std::move(other.pool);
+        other.count = 0;
+        return *this;
+    }
+
+    /**
+     * Creates a new object of typename T, by forwarding any and all arguments
+     * to the typename T constructor.
+     *
+     * @param args Arguments to be forwarded to the typename T constructor.
+     * @return A pointer to the newly created object, or nullptr on error.
+     */
+    template<typename... Args>
+    T* create(Args&&... args) {
+        int32_t capacity = pool.getCapacity();
+        if (count == capacity &&
+            pool.resize(2 * capacity, capacity) == nullptr) {
```

Comment:
Exponential growth is good, but you are starting from a very small capacity. Allocating a little bit more will be cheaper than regrowing a few times. Please grow faster for low values.

For example: `pool.resize(capacity < 10 ? 32 : 2 * capacity, capacity)`

Reply:
I added logging to the destructor, ran our unit tests (`make check`), counted how many `MemoryPool` objects had what `count` value when they got destructed and the top-10 then looked like this:

Amount|Count
-|-
21543|0
2288|6
528|3
275|2
136|5
134|1
31|18
10|9
10|24
4|7

Considering how common it is for `MemoryPool` objects to end up storing nothing at all (just being kept around in case something would need to be stored), I think the initial size ought to be as small as possible to not waste memory.

It's not possible to make a `MaybeStackArray` with smaller initial capacity than 1, so with irregular initial growth I think that 1 would be the sensible choice. Then doing an initial jump from 1 to 8 covers 99.7% of all the `MemoryPool` objects created, while giving a reasonably large starting size for those few that end up allocating a lot more objects.